### DISCUSSION TITLE
App secret keys?

### DISCUSSION THREAD

**Discussion by benallfree** [pocketbase/pocketbase#3587] [Q&A]
Is there any planned or existing support for something like a machine a secret app key aka API 'machine key' that doesn't expire? I'd like to be able to establish an admin client connection from a nodejs app and bypass the traditional login process and simply provide a secret key token instead.

**Comment by ganigeorgiev** [comment]
For now this is left aside as I have some security concerns with this approach (_you can check https://github.com/pocketbase/pocketbase/issues/3118, https://github.com/pocketbase/pocketbase/discussions/1314, https://github.com/pocketbase/pocketbase/discussions/342 and probably others_).

As a workaround you can create an admin account and use its email/pass credentials (_in the end there is no much difference compared to client id and secret_).
The JS SDK comes also with a builtin admin token refresh helper so it is fine to use small token duration for the admin token.

```js
await pb.admins.authWithPassword(ADMIN_EMAIL, ADMIN_PASSWORD, {
  autoRefreshThreshold: 5 * 60 // auto refresh if the token is expired or going to expire in the next 5 minutes
})
```

**Comment by benallfree** [comment]
Along these same lines... I will have a NodeJS app and PocketBase running inside the same Docker container. Can you think of a way using a pb_hook to automatically authenticate the connection as an admin role if it comes from localhost? This would avoid passing admin login/pw as environment variables.

**Reply by ganigeorgiev** [reply]
Hm, you can create a middleware and populate the admin context from there but note that if you are behind a reverse proxy you may need some other approach. For example (_treat the below as pseudo-code as I haven't tested it_):

```js
// NB! don't use if behind local reverse proxy because all requests will have the same remoteAddr
routerUse((next) => {
    return (c) => {
        // ip:port string (not sure if the below will work on ipv6 localhost format)
        const addr = c.request().remoteAddr;
        const ip = addr.substring(0, addr.lastIndexOf(":"));

        if (
            // has no explicit authorization header
            !c.request().header.get("authorization") &&
            // is localhost
            (ip === "localhost" || ip === "127.0.0.1" || ip === "[::1]")
        ) {
            // authenticate as test@example.com admin
            c.set("admin", $app.dao().findAdminByEmail("test@example.com"))
        }

        return next(c)
    }
})
```
